x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -411,12 +415,17 @@ public int adjustSampleCount(int originalCount) {
                     // the sum of the upscaled count values is less that totalCount.
                     // This code needs to be refactored to move all scaling into the server
                     // side, not just the resampling-scaling.
-                    return (int) Math.floor(newCount / (p));
+                    rawCount = (int) Math.floor(newCount / (p));
                 } else {
-                    return 0;
+                    rawCount = 0;
                 }
             } else {
-                return originalCount;
+                rawCount = originalCount;
+            }
+            if (adjustSampleCount) {
+                return (int) Math.floor(rawCount / sampleRate);
```

Comment:
The adjusted sample count will be undercounted here if there are two rounds of upscaling and both upscaling factors are greater than one. This will be more noticeable for larger time periods and/or larger datasets.

This happens because of the second use of `Math.floor`:

$$\lfloor count / p / sampleRate \rfloor \geq \lfloor \lfloor count / p \rfloor / sampleRate \rfloor$$

I think one possible suggestion here is to calculate the combined adjusted sample rate in the constructor and use it in `adjustSampleCount`. This would also allow you to remove the `adjustSampleCount` branch. I used this approach in a separate implementation a few weeks ago, but unfortunately I didn't measure the performance impact so I don't know how that will be impacted.

Reply:
Good catch! I've addressed this in https://github.com/elastic/elasticsearch/pull/99091/commits/9df6232ea0644a87124db46e6290e3fb2b3dc0d6 and also added tests. You'll notice that I left one TODO in the constructor of `Resampler`. That's a reminder for later, once we have changed all client code so we can do all resampling on the server-side.